src/eval.c

```diff
@@ -5394,6 +5465,774 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+    // string constants returns without string()
+    if (is_started_with_quote(embedded->expr))
+    {
+	char_u* already_stringified = disclose(embedded->expr);
+	embedded->stringified = already_stringified;
+	return;
+    }
+
+    // Apply string()
+    applying = alloc(STRLEN(embedded->expr) + STRLEN("string()") + 1);
+    applying_tofree = applying;
+
+    sprintf((char*) applying, "string(%s)", embedded->expr);
+    eval1(&applying, &applied, TRUE);
+    free(applying_tofree);
+
+    if (applied.v_type == VAR_UNKNOWN)
+    {
+	// Maybe, this VAR_UNKNOWN is occured only by 'Undefined variable'
+	semsg(_("E121: Undefined variable: %s"), embedded->expr);
+	embedded->stringified = NULL;
+	return;
+    }
+
+    // Assign this result
+    embedded->stringified = applied.vval.v_string;
+}
+
+/*
+ * Read all embedded from `template`,
+ * and check a passed template string is terminated.
+ *
+ * Return got an array of all embedded.
+ *
+ * This result MUST BE `free`d later! (if it is not NUL.)
+ */
+    static embedded_t**
+read_embedded(char_u* template, size_t* embedded_size)
+{
+    const size_t	LITERAL_EMBEDDING_START_LENGTH = 2;  // a number of "${"
+    const size_t	VAR_EMBEDDING_START_LENGTH = 1;  // a number of "$"
+    embedded_t**	embedded;
+    size_t		finished = 0;
+    char_u*		p;
+
+    *embedded_size = count_embedding(template);
+    embedded = (*embedded_size == 0)
+	? NULL  // Ignore the null allocation
+	: (embedded_t**) alloc(sizeof(embedded_t*) * *embedded_size);
+
+    for (p = template + 2; *p != NUL; MB_PTR_ADV(p))  // skip "$'"
+    {
+	if ((p[-1] == '$' && p[0] == '$') || (p[0] == '$' && p[1] == '$'))  // '$$' escapes to '$'
+	{
+	    continue;
+	}
+	else if (p[0] == '$')
+	{
+	    int embedded_is_literal = (p[1] == '{');
+	    int embedding_start_length;
+
+	    embedded[finished] = embedded_is_literal
+		? read_next_embedded_lit(p, template)
+		: read_next_embedded_var(p, template);
+	    if (embedded[finished] == NULL)
+	    {
+		free(embedded);
+		return NULL;
+	    }
+	    embedding_start_length =
+		(embedded_is_literal ? LITERAL_EMBEDDING_START_LENGTH : VAR_EMBEDDING_START_LENGTH)
+		+ STRLEN(embedded[finished]->expr);
+
+	    mb_ptr_adv_nth(&p, embedding_start_length);  // the part of "${literal" or "$var"
+	    ++finished;
+	    continue; // the part of "}" (increasing)
+	}
+    }
+
+    return embedded;
+}
+
+    static int
+read_next_string_length(char_u* whole_string, int expect_literal_string, char_u* template)
+{
+    char_u*	p;
+    int		(*is_closing_quote)(char_u*);
+
+    // If entered with the closing quote of a template string into this function
+    if (whole_string[1] == NUL)
+    {
+	semsg(_("E1000: Unterminated template literal: %s"), template);
+	return -1;
+    }
+
+    p = whole_string + 1;  // Skip the head of '"' (or '\'')
+    is_closing_quote = expect_literal_string
+	? is_closing_single_quote
+	: is_closing_double_quote;
+
+    for (; !is_closing_quote(p); MB_PTR_ADV(p))
+    {
+	switch (*p) {
```

Comment:
This `{` should be the next line.
Can we use `if (*p == NUL)`?

Reply:
!!
I fixed!